### PR TITLE
Constraint args

### PR DESCRIPTION
**Pull request by wpride**

### CODE REVIEW

**Comment by wpride** [comment]
@ctsims - what do you think of changing the spec to define constraints this way? arguments becoming evaluatable out of the box

**Comment by wpride** [comment]
could also do help/hint text this way

File: core/src/org/javarosa/xform/parse/XFormParser.java [hunk 1]
```diff
@@ -1030,6 +1032,32 @@ protected QuestionDef parseControl(IFormElement parent, Element e, int controlTy
         return question;
     }
 
+    private void parseConstraintLabel(QuestionDef q, Element e) {
```
Comment (wpride) [inline core/src/org/javarosa/xform/parse/XFormParser.java]: would be nice to reduce the repeated code between this and parseQuestionLabel (particularly if we use this pattern for help/hint text)

File: core/src/org/javarosa/form/api/FormEntryPrompt.java [hunk 2]
```diff
@@ -459,5 +459,17 @@ public void requestConstraintHint(ConstraintHint hint) throws UnpivotableExpress
...
+    /**
+     * Convenience method
+     * Get longText form of text for THIS element (if available)
+     * !!Falls back to default form if 'long' form does not exist.!!
+     * Use getSpecialFormQuestionText() if you want short form only.
+     *
+     * @return longText form
+     */
+    public String getConstraintText() {
+        return getQuestionText(getQuestion().getConstraintTextID());
```
Comment (wpride) [inline core/src/org/javarosa/form/api/FormEntryPrompt.java]: TODO: add fallback to previous method if this doesnt exist

**Comment by wpride** [comment]
@ctsims @orangejenny this is now working at least in my narrow test cases. I'm not sure about help images/audio/video at the moment - is there an example for this somewhere?

**Comment by orangejenny** [comment]
This app has hint text and a bunch of media in it: https://www.commcarehq.org/a/jennytraining/apps/view/ddd0b2eb8661eba592e59d9f7e9c0528/releases/

**Comment by wpride** [comment]
@orangejenny thanks. confirmed this works with help multimedia

**Comment by wpride** [comment]
https://confluence.dimagi.com/display/ccinternal/Constraint+Message+Spec+Change

@ctsims per our conversation yesterday
@orangejenny fyi

**Comment by ctsims** [comment]
Will - Ah, I see now. This does make sense to me and it definitely does
feel like it makes more sense for the label to be in the set of question
texts, given the MVC layout.

Couple comments:
1) we should still be namespacing the "constraint" here, since constraint
isn't in the XForms xml namespace.
2) We'll still need to continue supporting the old format of constraints as
well (we can make the parser load them in through an extra processing step
if need be). Is that supported in this PR directly? (jumping in soon).

-Clayton

**Comment by wpride** [comment]
2) fixed in new commit
can you elaborate on 1?

**Comment by ctsims** [comment]
Basically "constraint" isn't one of the fields in the xforms spec, so we
should theoretically be namespacing it into "our" xforms namespace

-Clayton

**Comment by ctsims** [comment]
Looks good pending tests. Should test:

A) That old constraint configuration still works reliably
B) That new constraint configuration works with the same (non-arg'd) setup as (A)
C) That new constraint configuration works with arguments

**Comment by wpride** [comment]
@ctsims

**Comment by ctsims** [comment]
Can you replace all "help"/"label" etc bare strings with a static final thing to make sure we don't make minor mistakes based on that?

**Comment by wpride** [comment]
@ctsims updated

**Pull request merged**